internal/persist/slice.go

```diff
@@ -0,0 +1,545 @@
+	// writeTail during t.Set is a race. Instead, we only set s.tail using
+	// t.setTail, which caches len(s.tail) in t.tailLen.
+	s       Slice[T]      // underlying slice data structure
+	aid     atomic.Uint64 // id of this transient (see [TransientSlice.id])
+	tailLen int           // cached copy of len(t.s.tail)
+	wtail   atomic.Bool   // whether s.tail is writable
+	tailMu  sync.Mutex    // protects tail copy in writeTail
+}
+
+// transientID is the most recently used transient ID.
+// To obtain a new ID, use transientID.Add(1).
+var transientID atomic.Uint64
+
+// id returns the id for this TransientSlice, copied from t.aid.
+// If an id has not been allocated yet (t.aid==0),
+// id sets t.aid to a new id and returns that id.
+func (t *TransientSlice[T]) id() uint64 {
+Retry:
+	id := t.aid.Load()
+	if id == 0 {
+		id = transientID.Add(1)
+		if !t.aid.CompareAndSwap(0, id) {
+			// Racing with another t.id call.
+			goto Retry
+		}
+	}
+	return id
+}
+
+// The tree uses chunks of size 16 and 16-way branching in the interior nodes.
+// A power of two is convenient.
+// Clojure uses 32, but BenchmarkAppend runs fastest with 16.
+const (
+	chunkBits = 4
+	chunkMask = chunk - 1
+	chunk     = 1 << chunkBits
+)
+
+// height returns the height of a tree with tlen elements.
+func height(tlen int) int {
+	if tlen == 0 {
+		return 0
+	}
+	return max(2, 1+bits.Len(uint(tlen-1))/chunkBits)
+}
+
+// A leaf is a leaf node in the tree.
+type leaf[T any] struct {
+	val [chunk]T // values
+	id  uint64   // id of TransientSlice that can write this leaf
+}
+
+// An inode is an interior node in the tree.
+type inode struct {
+	ptr [chunk]atomic.Value // all *leaf[T] or all *node, depending on tree level
+	id  uint64              // id of TransientSlice that can write this node
+}
+
+// Len returns len(s).
+func (s *Slice[T]) Len() int {
+	return s.tlen + len(s.tail)
+}
+
+// Len returns len(t).
+func (t *TransientSlice[T]) Len() int {
+	// Note: using t.tailLen to avoid race with t.writeTail overwriting t.s.tail
+	// (but not changing len(t.s.tail)).
+	return t.s.tlen + t.tailLen
+}
+
+// At returns s[i].
+func (s *Slice[T]) At(i int) T {
+	if i < 0 || i >= s.Len() {
+		panic(fmt.Sprintf("index %d out of range [0:%d]", i, s.Len()))
+	}
+	if i >= s.tlen {
+		return s.tail[i&chunkMask]
+	}
+	p := s.tree.Load()
+	for shift := (s.height - 1) * chunkBits; shift > 0 && p != nil; shift -= chunkBits {
+		p = p.(*inode).ptr[(i>>shift)&chunkMask].Load()
+	}
+	if p == nil {
+		var zero T
+		return zero
+	}
+	return p.(*leaf[T]).val[i&chunkMask]
+}
+
+// At returns t[i].
+func (t *TransientSlice[T]) At(i int) T { return t.s.At(i) }
+
+// All returns an iterator over s[0:len(s)].
+func (s *Slice[T]) All() iter.Seq2[int, T] {
+	return s.Slice(0, s.Len())
+}
+
+// All returns an iterator over t[0:len(t)].
+func (t *TransientSlice[T]) All() iter.Seq2[int, T] { return t.s.All() }
+
+// Slice returns an iterator over s[i:j].
+func (s *Slice[T]) Slice(i, j int) iter.Seq2[int, T] {
```

Comment:
this feels very expensive (see catenate for example). is there a way to group bulk access better for All?

Reply:
Most bulk access does use element-at-a-time loops, in which case the iterator is grouping the bulk access reasonably well by not rewalking the tree for each element. The entire iteration over n elements is O(n) instead of the O(n log n) that would result from `for i := range s.Len() { use s.At(i) }`.

I'm still not happy with the number of NewVectorSeq calls, especially NewVectorSeq(x.Slice(i, j)) for some x, i, j. I've thought about making Vector have at least an offset and length of its own, and then those can avoid a copy. Perhaps later.

Catenate is a different concern; replied below.